demo/style_configurator.c

```diff
@@ -0,0 +1,1048 @@
+// use the duplicate array method, or just let the user
+// manually set those after calling the function by accessing ctx->style->*?
+
+static const char* symbols[NK_SYMBOL_MAX] =
+{
+    "NONE",
+    "X",
+    "UNDERSCORE",
+    "CIRCLE_SOLID",
+    "CIRCLE_OUTLINE",
+    "RECT_SOLID",
+    "RECT_OUTLINE",
+    "TRIANGLE_UP",
+    "TRIANGLE_DOWN",
+    "TRIANGLE_LEFT",
+    "TRIANGLE_RIGHT",
+    "PLUS",
+    "MINUS"
+};
+
+
+
+static void
+style_rgb(struct nk_context* ctx, const char* name, struct nk_color* color)
+{
+	nk_label(ctx, name, NK_TEXT_LEFT);
+	if (nk_combo_begin_color(ctx, *color, nk_vec2(nk_widget_width(ctx), 400))) {
+		nk_layout_row_dynamic(ctx, 120, 1);
+		struct nk_colorf colorf = nk_color_picker(ctx, nk_color_cf(*color), NK_RGB);
+		nk_layout_row_dynamic(ctx, 25, 1);
+		colorf.r = nk_propertyf(ctx, "#R:", 0, colorf.r, 1.0f, 0.01f,0.005f);
+		colorf.g = nk_propertyf(ctx, "#G:", 0, colorf.g, 1.0f, 0.01f,0.005f);
+		colorf.b = nk_propertyf(ctx, "#B:", 0, colorf.b, 1.0f, 0.01f,0.005f);
+
+		*color = nk_rgb_cf(colorf);
+
+		nk_combo_end(ctx);
+	}
+}
+
+// TODO style_style_item?  how to handle images if at all?
+static void
+style_item_color(struct nk_context* ctx, const char* name, struct nk_style_item* item)
+{
+	style_rgb(ctx, name, &item->data.color);
+}
+
+static void
+style_vec2(struct nk_context* ctx, const char* name, struct nk_vec2* vec)
+{
+	char buffer[64];
+	nk_label(ctx, name, NK_TEXT_LEFT);
+	sprintf(buffer, "%.2f, %.2f", vec->x, vec->y);
+	if (nk_combo_begin_label(ctx, buffer, nk_vec2(200,200))) {
+		nk_layout_row_dynamic(ctx, 25, 1);
+		nk_property_float(ctx, "#X:", -100.0f, &vec->x, 100.0f, 1,0.5f);
+		nk_property_float(ctx, "#Y:", -100.0f, &vec->y, 100.0f, 1,0.5f);
+		nk_combo_end(ctx);
+	}
+}
+
+// style_general? pass array in instead of static?
+static void
+style_colors(struct nk_context* ctx, struct nk_color color_table[NK_COLOR_COUNT])
+{
+	const char* color_labels[NK_COLOR_COUNT] =
+	{
+		"COLOR_TEXT:",
+		"COLOR_WINDOW:",
+		"COLOR_HEADER:",
+		"COLOR_BORDER:",
+		"COLOR_BUTTON:",
+		"COLOR_BUTTON_HOVER:",
+		"COLOR_BUTTON_ACTIVE:",
+		"COLOR_TOGGLE:",
+		"COLOR_TOGGLE_HOVER:",
+		"COLOR_TOGGLE_CURSOR:",
+		"COLOR_SELECT:",
+		"COLOR_SELECT_ACTIVE:",
+		"COLOR_SLIDER:",
+		"COLOR_SLIDER_CURSOR:",
+		"COLOR_SLIDER_CURSOR_HOVER:",
+		"COLOR_SLIDER_CURSOR_ACTIVE:",
+		"COLOR_PROPERTY:",
+		"COLOR_EDIT:",
+		"COLOR_EDIT_CURSOR:",
+		"COLOR_COMBO:",
+		"COLOR_CHART:",
+		"COLOR_CHART_COLOR:",
+		"COLOR_CHART_COLOR_HIGHLIGHT:",
+		"COLOR_SCROLLBAR:",
+		"COLOR_SCROLLBAR_CURSOR:",
+		"COLOR_SCROLLBAR_CURSOR_HOVER:",
+		"COLOR_SCROLLBAR_CURSOR_ACTIVE:",
+		"COLOR_TAB_HEADER:"
+	};
+
+	nk_layout_row_dynamic(ctx, 30, 2);
+	for (int i=0; i<NK_COLOR_COUNT; ++i) {
+		style_rgb(ctx, color_labels[i], &color_table[i]);
+	}
+
+	nk_style_from_table(ctx, color_table);
+}
+
+static void
+style_text(struct nk_context* ctx, struct nk_style_text* out_style)
+{
+	struct nk_style_text text = *out_style;
```

Comment:
Yeah I agree the pointer is better, though I'll definitely change the name of the parameters to match the current local style variables.  I think my reasoning at the time was (as usual) laziness, and knowing I'd probably change it later.  Though I honestly don't think it'd make a noticeable performance difference unless you were using unoptimized tcc builds on some low powered ARM core or something.  I'm continually amazed at how fast even Raspberry Pi's etc. are these days.

Wait isn't your first example wrong?  That's why I was too lazy to do it at first?  Shouldn't it be

    static void
    style_text(struct nk_context* ctx, struct nk_style_text* out_style)
    {
        nk_layout_row_dynamic(ctx, 30, 2);
        style_rgb(ctx, "Color:", &out_style->color);
        style_vec2(ctx, "Padding:", &out_style->padding);
    }

So my final version will be

    static void
    style_text(struct nk_context* ctx, struct nk_style_text* text)
    {
        nk_layout_row_dynamic(ctx, 30, 2);
        style_rgb(ctx, "Color:", &text->color);
        style_vec2(ctx, "Padding:", &text->padding);
    }

Reply:
Yep, my example was incorrect. Didn't actually compile them :)

Reply:
Np, you can tell I didn't even notice till I was halfway through typing my response.

I've often heard/thought that the . vs -> is the one thing thing that the compiler could actually just issue a warning and fix automatically without any potential downsides.  Or get rid of -> entirely but I disagree with that since I like the visual reminder something is a pointer.

And clearly the overhead of typing an extra character is enough that I often do this copying to a local variable thing.